components/suggest/src/suggest.udl

```diff
@@ -8,35 +8,46 @@ typedef extern RemoteSettingsConfig;
 
 namespace suggest {
 
+boolean suggestion_url_matches([ByRef] SuggestionUrl url, [ByRef] string other_url_string);
+
 };
 
 [Error]
 interface SuggestApiError {
     Other(string reason);
 };
 
-
 enum SuggestionProvider {
     "Amp",
     "Wikipedia",
 };
 
+dictionary SuggestionUrl {
+    string string;
+    sequence<SuggestionUrlReplacement> replacements;
+};
```

Comment:
This interface is a little awkward because of mozilla/uniffi-rs#1372. At first, I modeled `SuggestionUrl` as an object, like:

```webidl
interface SuggestionUrl {
    string string();
    boolean matches([ByRef] string other_url_string);
};
```

But `Suggestion` is an enum with associated data, and those can't have object fields yet. Even if they could, though, `SuggestionUrl` would need an explicit destructor in Kotlin (mozilla/uniffi-rs#1394).

Another idea I considered was turning `Suggestion` into an "enum-like" object with optional-returning getters for each variant, like this:

```webidl
interface Suggestion {
    AmpSuggestion? as_amp_suggestion();
    WikipediaSuggestion? as_wikipedia_suggestion();
};

dictionary AmpSuggestion {
    string title;
    SuggestionUrl url;
    sequence<u8>? icon;
    // ...
};

dictionary WikipediaSuggestion {
    string title;
    SuggestionUrl url;
    // ...
};
```

But that makes `Suggestion` feel clunkier, and forces explicit destructors for `Suggestion` _and_ `SuggestionUrl` in Kotlin.

Also, UniFFI dictionaries can't have methods, so that's why `suggestion_url_matches` is a free function instead of a method on `SuggestionUrl`.

/cc @bendk @mhammond Do you folks have any suggestions (😅) for other things to try, or does this interface look okay to you, given UniFFI's restrictions on it?

Reply:
I can't really think of anything better.  This makes me want to address 1372 and let users configure if they want to treat their interface as a resource or not.